### PR TITLE
HHH-12454 - Offer flag to consider id generator with local scope (legacy non JPA behavior)

### PR DESCRIPTION
**Pull request by dreab8**
https://hibernate.atlassian.net/browse/HHH-12454

@sebersole any idea for a better name than ID_GENERATOR_NAMES_GLOBAL_SCOPE = "hibernate.id.generator.global_scope_name";?

Backporting to 5.2 do we have to keep the default value to true?

### CODE REVIEW

**Comment by sebersole** [comment]
`hibernate.id.generator.allow_illogical_jpa_global_definitions`? ;)

**Comment by vladmihalcea** [comment]
@dreab8 Maybe `hibernate.id.generator.enable_legacy_entity_scope`?

**Comment by sebersole** [comment]
`hibernate.id.generator.global_scoping_enabled`

**Comment by dreab8** [comment]
hibernate.id.generator.global_scoping_enabled sound good to me
Thanks @sebersole and @vladmihalcea

**Comment by sebersole** [comment]
Shame.. I really liked `allow_illogical_jpa_global_definitions` :)

**Comment by vladmihalcea** [comment]
@dreab8 @sebersole If we are to backport this to 5.2, maybe we should use the `false` value by default.

From 5.3, we should have it enabled by default, but users will pay more attention to migration guides when upgrading to a major/minor version, than when they migrate to a patch verson.

What do you think?

File: hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java [hunk 1]
```diff
@@ -760,7 +762,13 @@ else if ( InheritanceType.SINGLE_TABLE.equals( inheritanceState.getType() ) ) {
 		}
 
 		// try to find class level generators
-		HashMap<String, IdentifierGeneratorDefinition> classGenerators = buildGenerators( clazzToProcess, context );
+		HashMap<String, IdentifierGeneratorDefinition> classGenerators;
+		if ( context.getBuildingOptions().isIdGeneratorNameScopeGlobal() ) {
```
Comment (vladmihalcea) [inline hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java]: Here, there is no difference between the if and else branches. Did you have something in mind and forgot to add it?

**Comment by sebersole** [comment]
> If we are to backport this to 5.2, maybe we should use the false value by default.

This is always tricky.  On the one hand is JPA compatibility and the other is upgradeability (legacy expectations).  TBH I am not exactly sure of the legacy expectations here.  @emmanuelbernard  probably knows better since he wrote all of the annotation binding/processing code.

I also think it is bad form to change these default values from release to release.  Whatever default we pick for 5.2 ought to the be the default moving forward.  IMO

File: hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java (hunk 1 above)
Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/cfg/AnnotationBinder.java]: good catch, Thanks @vladmihalcea

**Comment by dreab8** [comment]
@sebersole I'm torn but probably it is better to keep true as a default value also for 5.2.
@vladmihalcea I'm going to create a PR also for 5.2

**Comment by sebersole** [comment]
Actually, I change my mind on this if I still can... this is really nothing more that a jpa compliance this.  So I think this should be one of those `"hibernate.jpa.compliance.<>"` / `JpaCompliance` things.  

`"hibernate.jpa.compliance.global_id_generators"`?

**Comment by dreab8** [comment]
@sebersole in order to make `JpaCompliance `available to the `AnnotationBinder`, is it fine to add it to the `BootStrapContext`?
Currently the `JpaComliance` is available only through `SessionFactoryOptions`

**Comment by sebersole** [comment]
Well, first do you agree that modeling it as a JpaCompliance setting is the correct/logical/natural thing to do?  

If not, no worries.

If so, yes, we will need to make it available during the earlier bootstrap phases - which is not a bad thing.  I do worry about the `JpaCompliance` object being mutated after its value(s) have been used.  E.g. if the user is allowed to change `JpaCompliance#isGlobalGeneratorScopeEnabled` (proposed name) after we have fully built the Metadata object - that would have a surprising outcome (no effect).  We just need to make sure that we properly "lock" the `JpaCompliance` at the point of Metadata -> SessionFactoryBuilder

**Comment by dreab8** [comment]
I think the property is JPA compliance related  and I agree also with not allowing the user to change this property . 
Not sure but probably a simple solution for the immutability can be to initialise the property in the JpaComplianceImpl constructor and do not provide a setter method for it, what do you think?

**Comment by sebersole** [comment]
Well its the same for the others.  They are all accessed via setters that
are just not exposed from `JpaCompliance` (only `JpaComplianceImpl`).  But
they are still there - even at runtime.  If we move this thing to
`BootstrapContext` then a user's only option to mutate these is through
some sort of reference to `JpaCompliance` (or `MutableJpaCompliance`). I'd
really hate for the "API" calls to be `( (JpaComplianceImpl)
metadataBuilder.getBootstrapContext().getJpaCompliance()
).setSomething(..)` - that's fugly.  Much better as something like just `
metadataBuilder.getBootstrapContext().getJpaCompliance().setSomething()`
which means that whatever is returned from
`BootstrapContext.getJpaCompliance()`
needs to be a mutable view.

**Comment by dreab8** [comment]
For letting users to mutate these properties I was thinking to leave the `SessionFactoryOptionBuilder#enableJpaQueryCompliance` and the others methods you added.

The `BootstrapContext` build and return a `JpaCompliance` that does not provide any setter. 
The `SessionFactoryOptionBuilder` gives the user a way to mutate the properties values.

`JpaCompliance` can be an spi interface, but I may be wrong.

**Comment by dreab8** [comment]
@sebersole   after thinking a little more :) , I realized your idea of having a mutable and immutable interface is  good, I'll work on this idea and make another PR.

Thanks

**Comment by dreab8** [comment]
@sebersole tried to follow your suggestions, let me know what do u think.
Thanks

**Comment by sebersole** [comment]
@dreab8 I see 2 PRs linked to this Jira - this one, and https://github.com/hibernate/hibernate-orm/pull/2232.  I assume one is out-of-date?  Anything else you are waiting on?

**Comment by dreab8** [comment]
@sebersole can you give a look at this [last commit](https://github.com/hibernate/hibernate-orm/pull/2228/commits/a5fa173491ca9b98b5b01dc678203ef7eb1f78b1), it's related with JpaCompliance immutability.
The other [PR](https://github.com/hibernate/hibernate-orm/pull/2232) is for the 5.2 branch

**Comment by sebersole** [comment]
Sure.. but first, do you think I'm being overly paranoid about the mutability?

**Comment by dreab8** [comment]
@sebersole no I don't think you are :)

File: hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java [hunk 2]
```diff
@@ -975,7 +975,7 @@ public boolean isFailOnPaginationOverCollectionFetchEnabled() {
 
 	@Override
 	public JpaCompliance getJpaCompliance() {
-		return jpaCompliance;
+		return jpaCompliance.immutableCopy();
```
Comment (sebersole) [inline hibernate-core/src/main/java/org/hibernate/boot/internal/SessionFactoryOptionsBuilder.java]: Is this only ever called once?  If not, maybe we want to change this so we do only call it once.  Doubt it has a huge perf impact, but every bit helps

**Comment by sebersole** [comment]
Applied upstream

**Comment by icyerasor** [comment]
Found the property name _global_id_generators_ really confusing today; as to me it implied that the generators themselves would be shared globally, not only their definitions / names.
https://forum.hibernate.org/viewtopic.php?p=2338260 and https://vladmihalcea.com/hibernate-hidden-gem-the-pooled-lo-optimizer/ finally helped me to grasp the concept I guess (hope) :D

**Comment by FireRedDev** [comment]
Hey @icyerasor ! Does this issue's flag fix problems with higher Hibernate versions with specifying @SequenceGenerator for extending a base entity with an id as mentioned in the comments in the top-answer here: https://stackoverflow.com/questions/8589928/mappedsuperclass-change-sequencegenerator-in-subclass

**Comment by icyerasor** [comment]
Which comment are you referring to? The one that it doesn't work anymore with hibernate 5.4 ([here](https://stackoverflow.com/questions/8589928/mappedsuperclass-change-sequencegenerator-in-subclass#comment130111215_12151985))?
It does still work for me with hibernate 5.6.10. Seems it was only broken for a few minor versions, see info ([here](https://stackoverflow.com/questions/4560813/specifying-distinct-sequence-per-table-in-hibernate-on-subclasses#comment92100438_4561233) and the links [1](https://hibernate.atlassian.net/browse/HHH-12329), [2](https://in.relation.to/2018/04/26/hibernate-orm-5217-final-release/)).

Warning is logged as expected: `HHH000069: Duplicate generator name`.
Haven't tried 6+ yet though.

**Pull request closed without merging**